src/base/rss/rss_autodownloader.cpp

```diff
@@ -0,0 +1,374 @@
+}
+
+void AutoDownloader::handleTorrentDownloadFinished(const QString &url)
+{
+    auto job = m_waitingJobs.take(url);
+
+    auto feed = Session::instance()->feedByURL(job->feedURL);
+    if (feed) {
+        if (auto article = feed->articleByGUID(job->articleGUID))
+            article->markAsRead();
+    }
+}
+
+void AutoDownloader::handleTorrentDownloadFailed(const QString &url)
+{
+    m_waitingJobs.remove(url);
+    // TODO: Re-schedule job here.
+}
+
+void AutoDownloader::handleNewArticle(Article *article)
+{
+    if (!article->isRead() && !article->torrentUrl().isEmpty())
+        addJobForArticle(article);
+}
+
+void AutoDownloader::setRule_impl(const AutoDownloadRule &rule)
+{
+    m_rules.insert(rule.name(), rule);
+}
+
+void AutoDownloader::addJobForArticle(Article *article)
+{
+    QSharedPointer<ProcessingJob> job(new ProcessingJob);
+    job->feedURL = article->feed()->url();
+    job->articleGUID = article->guid();
+    job->articleTitle = article->title();
+    job->torrentURL = article->torrentUrl();
+    m_processingQueue.append(job);
+    if (!m_processingTimer->isActive())
+        m_processingTimer->start();
+}
+
+void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
+{
+    foreach (const QString &ruleName, m_rules.keys()) {
+        AutoDownloadRule &rule = m_rules[ruleName];
+        if (!rule.isEnabled()) continue;
+        if (!rule.feedURLs().contains(job->feedURL)) continue;
+        if (!rule.matches(job->articleTitle)) continue;
+
+//        rule.setLastMatch(QDateTime::currentDateTime());
+//        m_dirty = true;
+
+        // if rule is in ignoring state do nothing with matched torrent
+        if (rule.ignoreDays() > 0) {
+            // Optimization: update lastMatch only for rules which use it for "ignoring" purposes
+            const QDateTime lastMatch = rule.lastMatch();
+            rule.setLastMatch(QDateTime::currentDateTime());
+            m_dirty = true;
+            if (lastMatch.isValid()) {
+                if (QDateTime::currentDateTime() < lastMatch.addDays(rule.ignoreDays()))
+                    return;
+            }
+        }
+
+        BitTorrent::AddTorrentParams params;
+        params.savePath = rule.savePath();
+        params.category = rule.assignedCategory();
+        params.addPaused = rule.addPaused();
+        BitTorrent::Session::instance()->addTorrent(job->torrentURL, params);
+
+        if (BitTorrent::MagnetUri(job->torrentURL).isValid()) {
+            auto feed = Session::instance()->feedByURL(job->feedURL);
+            if (feed) {
+                if (auto article = feed->articleByGUID(job->articleGUID))
+                    article->markAsRead();
+            }
+        }
+        else {
+            // waiting for torrent file downloading
+            m_waitingJobs.insert(job->torrentURL, job);
+        }
+
+        return;
+    }
+}
+
+void AutoDownloader::load()
+{  
+    QFile rulesFile(m_fileStorage->storageDir().absoluteFilePath(RulesFile));
```

Comment:
Consider renaming `RulesFile` into `RulesFileName`, because now this line looks confusing.